File: sdks/python/examples/webhooks.py
"""Example: Setting up webhooks with OpenAgentMail."""

import os

from openagentmail import OpenAgentMail

# Initialize client
client = OpenAgentMail(api_key=os.environ["OAM_API_KEY"])

# Create a webhook to receive message events
webhook = client.webhooks.create(
    url="https://your-server.com/webhooks/email",
    event_types=[
        "message.received",
        "message.sent",
        "message.bounced",
    ],
    client_id="main-webhook",  # For idempotency
)

print(f"Created webhook: {webhook.webhook_id}")
print(f"URL: {webhook.url}")
print(f"Events: {webhook.event_types}")
print(f"Secret: {webhook.secret}")  # Use this to verify webhook signatures

# Create a webhook for a specific inbox only
inbox = client.inboxes.create(username="monitored", client_id="monitored-inbox")

inbox_webhook = client.webhooks.create(
    url="https://your-server.com/webhooks/inbox-events",
    event_types=["message.received"],
    inbox_ids=[inbox.inbox_id],
)

print(f"\nCreated inbox-specific webhook: {inbox_webhook.webhook_id}")

# List all webhooks
print("\nAll webhooks:")
for wh in client.webhooks.list():
    status = "enabled" if wh.enabled else "disabled"
    print(f"  - {wh.url} ({status})")
    print(f"    Events: {wh.event_types}")

# Disable a webhook
disabled_webhook = client.webhooks.update(
    webhook.webhook_id,
    enabled=False,
)
print(f"\nDisabled webhook: {disabled_webhook.enabled}")

# Rotate webhook secret (in case of compromise)
rotated = client.webhooks.rotate_secret(webhook.webhook_id)
print(f"New secret: {rotated.secret}")

File: sdks/typescript/examples/webhooks.ts
// Example: Setting up webhooks

import { OpenAgentMail } from 'openagentmail';

async function main() {
  const client = new OpenAgentMail({
    apiKey: process.env.OAM_API_KEY!,
  });

  // Create a webhook for new messages
  const webhook = await client.webhooks.create({
    url: 'https://your-server.com/webhooks/email',
    eventTypes: ['message.received', 'message.sent'],
    // Optionally filter to specific inboxes
    // inboxIds: ['inbox_abc123'],
  });

  console.log('Created webhook:', webhook.webhookId);
  console.log('Webhook secret:', webhook.secret);
  console.log('Events:', webhook.eventTypes.join(', '));

  // Update webhook to disable it temporarily
  const updated = await client.webhooks.update(webhook.webhookId, {
    enabled: false,
  });
  console.log('\nWebhook enabled:', updated.enabled);

  // Rotate the webhook secret
  const rotated = await client.webhooks.rotateSecret(webhook.webhookId);
  console.log('\nNew secret:', rotated.secret);

  // List all webhooks
  console.log('\nAll webhooks:');
  for await (const wh of client.webhooks.list()) {
    console.log(`  - ${wh.url} (${wh.enabled ? 'enabled' : 'disabled'})`);
  }

  // Clean up
  await client.webhooks.delete(webhook.webhookId);
  console.log('\nDeleted webhook');
}

main().catch(console.error);
